Ensure vtable entries for generics are attached constants

dwblaikie opened the pull request:
Otherwise these end up as unattached constants (see the baseline test
changes) and can't be resolved by `GetConstantValueInSpecific` in
lowering or in further derived vtables.

If the class is non-generic, then it's fine for the vtable entry for
some function inherited from a generic base is represented as an
unattached constant, since the specific in that specific_function is
already fully resolved.

dwblaikie committed:
Baseline test

dwblaikie committed:
Ensure vtable entries for generics are attached constants

Otherwise these end up as unattached constants (see the baseline test
changes) and can't be resolved by `GetConstantValueInSpecific` in
lowering or in further derived vtables.

If the class is non-generic, then it's fine for the vtable entry for
some function inherited from a generic base is represented as an
unattached constant, since the specific in that specific_function is
already fully resolved.

danakj commented on toolchain/check/class.cpp:
```diff
@@ -201,6 +201,17 @@ static auto BuildVtable(Context& context, Parse::ClassDefinitionId node_id,
+                     context.sem_ir().insts().TryGetAs<SemIR::SpecificFunction>(
+                         base_vtable_entry_id)) {
+        if (class_generic_id.has_value()) {
+          derived_vtable_entry_id = GetOrAddInst<SemIR::SpecificFunction>(
```
Can we add a comment saying were creating a new instruction with the same values so that it will be a dependent instruction of the class generic, unlike the incoming `derived_vtable_entry_id`? 

Assuming my understanding of the change is correct?

dwblaikie replied:
Ah, yeah, for sure - 53a39b9c779c33cefe66e35e1a74bd74006e95f5

danakj commented on toolchain/check/class.cpp:
```diff
@@ -201,6 +201,17 @@ static auto BuildVtable(Context& context, Parse::ClassDefinitionId node_id,
         derived_vtable_entry_id = build_specific_function(*i);
         override_fn.virtual_index = vtable.size();
         CARBON_CHECK(override_fn.virtual_index == fn.virtual_index);
+      } else if (auto base_vtable_entry =
```
```suggestion
      } else if (auto base_vtable_specific_function =
```

danakj commented on toolchain/check/class.cpp:
```diff
@@ -201,6 +201,17 @@ static auto BuildVtable(Context& context, Parse::ClassDefinitionId node_id,
+      } else if (auto base_vtable_entry =
+                     context.sem_ir().insts().TryGetAs<SemIR::SpecificFunction>(
+                         base_vtable_entry_id)) {
+        if (class_generic_id.has_value()) {
```
If we drop the if condition here, does the semir change at all? GetOrAddInst seems like it would just return the existing constant value SpecificFunction if it could?

dwblaikie replied:
Per open discussion, seems this can't be dropped, since `GetOrAddInst` doesn't fully canonicalize, so removing this would add an extra instruction to the derived class.

But I did rephrase this check in 955ece7f5a2cac8d3c1a42643faf1788fa416813 to inspect the symbolicness of the ConstantValueInSpecific to determine whether this needed the attached constant or we could depend on the unattached/fully canonicalized constant instead.

dwblaikie committed:
Rename variable more precisely

Co-authored-by: Dana Jansens <danakj@orodu.net>

dwblaikie committed:
Updates for variable rename

dwblaikie committed:
Fix derived non-overload specific id

Because this referenced the original base function instead of the one
with the specific from the derived's base declaration, the derived
vtable entry still referenced the base's self specific instead of the
derived's base declaration's specific.

dwblaikie committed:
Add test coverage for a derived class in a generic context while not being explicitly generic itself.

dwblaikie committed:
Rephrase check for derived class in a generic context to use the symbolic/concreteness of the constant, instead of whether the class has a generic_id

This seems more like a direct expression of the intent/more legible,
even if it involves a little more plumbing/data to be passed around.

dwblaikie committed:
Add comment

danakj commented on toolchain/sem_ir/function.h:
```diff
@@ -192,8 +192,10 @@ auto GetCalleeFunction(const File& sem_ir, InstId callee_id,
 
 struct DecomposedVirtualFunction {
-  // The underlying `FunctionDecl`.
+  // The canonical instruction.
```
```suggestion
  // The canonical instruction from the `fn_decl_const_id`.
```

danakj reviewed (approved):
LGTM

dwblaikie committed:
clarify comment

Co-authored-by: Dana Jansens <danakj@orodu.net>
